### PR TITLE
Changing impl of keeping sealed index segs in mem

### PR DESCRIPTION
**Pull request by vgkholla**
After a discussion with @cgtz, impl has been changed to hold the serialized version of sealed index segments in memory instead of converting into a hash map.
This brings the benefit of reducing the memory footprint and making it more predictable but removes the ability to mutate sealed index  segments if required in the future

### CODE REVIEW

**Comment by pnarayanan** [comment]
Have you looked at `MappedByteBuffer#load()` option to load all pages into memory (as opposed to explicitly reading everything into JVM memory)? I think it can keep the code simple (one line code difference?). Additionally, exploiting native memory may have its benefits, and wouldn't require any changes to JVM memory arguments which we may have to otherwise tune. When used with swappiness = 0 on Linux, I would think that the pages will remain in memory after they get loaded, unless the system runs out of memory. This might also be a better choice in open source (no tuning required for JVM heap size and such).

**Comment by vgkholla** [comment]
@pnarayanan I took at look at `load()` and it says it "makes a best effort to ensure that, when it returns, this buffer's content is resident in physical memory" and does not say anything about whether it will _continue_ to be in memory. I'll chat with you tomorrow.

**Comment by pnarayanan** [comment]
which is what we want I believe - as long as you have enough memory, the best effort will succeed (if you look at the implementation, it just accesses an address in every page in the mapped file, to get the page loaded into memory).
With swappiness=0 setting (which is the setting on storage nodes we are familiar with afaik), pages loaded into memory do not get swapped out unless the system is running low on memory, which is what we want I think.

**Comment by cgtz** [comment]
I wonder if it would be worth having both options available via configs: `MappedByteBuffer::load` and explicitly loading into JVM heap.
We could try making a metric based on `MappedByteBuffer.isLoaded` to see if "loaded" mmap buffers get evicted from memory.

**Comment by vgkholla** [comment]
@pnarayanan @cgtz Addressed comments. Enhanced the config to provide three options
1) not have the index in mem
2) have the index in heap mem
3) force load the mmap into mem

File: ambry-api/src/main/java/com.github.ambry/config/StoreConfig.java [hunk 1]
```diff
@@ -248,12 +251,12 @@
   public static final String storeTtlUpdateBufferTimeSecondsName = "store.ttl.update.buffer.time.seconds";
 
   /**
-   * Specifies whether all index segments should be kept in memory (as opposed to a specific subset).
+   * Provides a hint for how indexes should be treated w.r.t memory
    */
-  @Config(storeKeepIndexInMemoryName)
+  @Config(storeIndexMemStateName)
   @Default("false")
```
Comment (pnarayanan) [inline ambry-api/src/main/java/com.github.ambry/config/StoreConfig.java]: you need to change this.

Comment (zzmao) [reply ambry-api/src/main/java/com.github.ambry/config/StoreConfig.java]: @vgkholla  Can you change the default value? Then I can merge this change.

File: ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java [hunk 2]
```diff
@@ -702,59 +696,69 @@ void writeIndexSegmentToFile(Offset safeEndPoint) throws IOException, StoreExcep
...
   }
 
   /**
-   * Memory maps the segment of index.
+   * Maps the segment of index either as a memory map or a in memory buffer depending on config.
    * @throws IOException if there is any problem reading or mapping files
    * @throws StoreException if there are problems with the index
    */
-  private void mmap() throws IOException, StoreException {
+  private void map() throws IOException, StoreException {
```
Comment (zzmao) [inline ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java]: Looks like `mapped` is set to true only when `sealed` is set to true. Can we just use one variable?

**Comment by zzmao** [comment]
@vgkholla  Two minor comments. Then we can merge this PR.

File: ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java (hunk 2 above)
Comment (vgkholla) [reply ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java]: That's right, they are the same with this change

File: ambry-api/src/main/java/com.github.ambry/config/StoreConfig.java (hunk 1 above)
Comment (vgkholla) [reply ambry-api/src/main/java/com.github.ambry/config/StoreConfig.java]: Fixed

**Comment by vgkholla** [comment]
@zzmao addressed comments

**Pull request merged**